### DISCUSSION TITLE
Problem with webhook

### DISCUSSION THREAD

**Discussion by gapnet** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#30] [General]
Hi, first thanks for your great job. I'm having a problem with registering the webhook. The verify token is correct, I'm trying with https://a096-186-137-155-57.sa.ngrok.io/api/WhatsAppNotification/webhooks, even changed to [HttpGet("webhooks")] instead of [HttpGet("receive/TextMessage")] in ConfigureWhatsAppMessageWebhook, but can't get facebook platform to validate it. Can you please tell me what i'm doing wrong? Thanks in advance.

**Comment by gabrieldwight** [comment]
Hi, for registering the webhook subscription.  Make sure you return the hubChallenge result response which will successfully validate your webhook.

```C#
private string VerifyToken = "<YOUR VERIFY TOKEN STRING>";

[HttpGet("webhooks")]
        public ActionResult<string> ConfigureWhatsAppMessageWebhook([FromQuery(Name = "hub.mode")] string hubMode,
                                                                    [FromQuery(Name = "hub.challenge")] int hubChallenge,
                                                                    [FromQuery(Name = "hub.verify_token")] string hubVerifyToken)
{
_logger.LogInformation("Results Returned from WhatsApp Server\n");
            _logger.LogInformation($"hub_mode={hubMode}\n");
            _logger.LogInformation($"hub_challenge={hubChallenge}\n");
            _logger.LogInformation($"hub_verify_token={hubVerifyToken}\n");

            if (!hubVerifyToken.Equals(VerifyToken))
            {
                return Forbid("VerifyToken doesn't match");
            }
            return Ok(hubChallenge);
}
```

**Reply by gapnet** [reply]
Thanks for your response!. Yes, I'm returning Ok(hubChallenge), but facebook is not even getting to the webhook EndPoint, 

https://a096-186-137-155-57.sa.ngrok.io/api/WhatsAppNotification/webhooks

I think I'm doing something wrong with the URL but can't figure what is. I even tried a webhook with glitch and didn't have problems.

![03-02-2023 05-06-21 p m](https://user-images.githubusercontent.com/12832607/216700698-a6cf8c0b-a2ea-4c64-8c19-5efc1b249441.png)

**Reply by gabrieldwight** [reply]
The problem comes from ngrok local tunnel. Sometimes ngrok usually cause some delays and make the webhook not reachable. If it is not reachable you can try with dev tunnel functionality from visual studio 2022 or you can install conveyor extension to simulate the webhook.

**Reply by gapnet** [reply]
Thanks for your answer. The problem was ngrok. Changed to dev tunnel and saved the webhook. Thanks again for your time.

**Comment by rodrigoramirez93** [comment]
Hey guys, I'm hitting the same issue. I'm struggling to understand why I get this same issue:

![imagen](https://user-images.githubusercontent.com/19152911/230706878-d02b7c3d-e1d4-477e-98f6-d2df30e988c4.png)

In my case, I have a deployed app service and I just simply opened the two endpoints /webhooks POST and GET. With glitch I don't have this problem. Any idea what can be wrong?

**Reply by gabrieldwight** [reply]
Sometimes ngrok usually has an issue receiving a webhook request from Meta. If you are testing locally you can use dev tunnel feature from visual studio 2022 to test your webhook.

**Comment by siglaadmprojetos** [comment]
Hi @gabrieldwight , first of all, thanks for your great job!
I'm facing a similar problem to the colleague, but when it comes to receiving and replying to messages. I'm managing to receive the json in the webhook when the user starts the conversation, but I'm not able to respond.

![Captura de tela 2023-04-23 234425](https://user-images.githubusercontent.com/101640996/233891590-f2a9ad84-2a59-4de8-80d0-cb028d1cf7b1.png)

I can receive and save the json in the bank. But when retrieving it, I use the same method to reply. Below is the snippet



![Captura de tela 2023-04-23 235730](https://user-images.githubusercontent.com/101640996/233893688-641c9439-0a30-4e8a-bd0f-4ff2982ceb52.png)

in use Try...catch to show a error and save  it:


![Captura de tela 2023-04-24 003003](https://user-images.githubusercontent.com/101640996/233893940-bfa48652-03d9-4abf-a5e8-02a49b8d345d.png)

It´s about error in line 344, saying that not exists definition for 'SingleOrDefault'. How can I deal with this error?
Thanks for the help.

**Reply by gabrieldwight** [reply]
Hi @siglaadmprojetos, the error for SingleOrDefault means the dynamic object of the **messageReceived** needs to be cast as an array or a collection which will make the **SingleOrDefault** method to be accessible.

The JSON you receive from WhatsApp:
```JSON
{
  "object": "whatsapp_business_account",
  "entry": [{
      "id": "WHATSAPP_BUSINESS_ACCOUNT_ID",
      "changes": [{
          "value": {
              "messaging_product": "whatsapp",
              "metadata": {
                  "display_phone_number": PHONE_NUMBER,
                  "phone_number_id": PHONE_NUMBER_ID
              },
              "contacts": [{
                  "profile": {
                    "name": "NAME"
                  },
                  "wa_id": PHONE_NUMBER
                }],
              "messages": [{
                  "from": PHONE_NUMBER,
                  "id": "wamid.ID",
                  "timestamp": TIMESTAMP,
                  "text": {
                    "body": "MESSAGE_BODY"
                  },
                  "type": "text"
                }]
          },
          "field": "messages"
        }]
  }]
}
```

You can solve it like this:
```C#
private List</*Your class Name*/> textMessage;
// inside on your method
textMessage = new List</*Your class Name*/>(messageReceived.Entry.SelectMany(x => x.Changes).SelectMany(x => x.Value.Messages));

// accessing the message properties
string messageId = textMessage.SingleOrDefault().Id;
string sender = textMessage.SingleOrDefault().From;

// Then you can send your reply meesage
```
